pkgs/native_assets_cli/lib/src/c_compiler_config.dart

```diff
@@ -0,0 +1,129 @@
+// Copyright (c) 2023, the Dart project authors.  Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+import 'package:collection/collection.dart';
+
+import 'json_utils.dart';
+import 'utils/map.dart';
+
+/// The configuration for a C toolchain.
+final class CCompilerConfig {
+  /// Path to a C compiler.
+  late final Uri? compiler;
+
+  /// Path to a native linker.
+  late final Uri? linker;
+
+  /// Path to a native archiver.
+  late final Uri? archiver;
+
+  /// Path to script that sets environment variables for [compiler], [linker],
+  /// and [archiver].
+  late final Uri? envScript;
+
+  /// Arguments for [envScript].
+  late final List<String>? envScriptArgs;
+
+  /// Constructs a new [CCompilerConfig] based on the given toolchain tools.
+  CCompilerConfig({
+    this.archiver,
+    this.compiler,
+    this.linker,
+    this.envScript,
+    this.envScriptArgs,
+  });
+
+  /// Constructs a [CCompilerConfig] from the given [json].
+  ///
+  /// The json is expected to be valid encoding obtained via
+  /// [CCompilerConfig.toJson].
+  factory CCompilerConfig.fromJson(Map<String, Object?> json) {
+    final compiler = _parseCompiler(json);
+    return CCompilerConfig(
+      archiver: _parseArchiver(json),
+      compiler: compiler,
+      envScript: _parseEnvScript(json, compiler),
+      envScriptArgs: _parseEnvScriptArgs(json),
+      linker: _parseLinker(json),
+    );
+  }
+
+  // TODO(https://github.com/dart-lang/native/issues/1599): The main reason
+  // these keys are exposed is due to usage on Dart CI testing infrastructure.
+  // If our infrastructure supplies the native toolchains via passing those down
+  // to the `dart build/run` via overrides we should no longer need to expose
+  // these here.
+  static const configKey = 'c_compiler';
```

Comment:
Can these be consts in an extension and only exposed in `internal.dart`? Or can we just inline them in the Dart CI and make them private here? (at the cost of possibly having them get out of sync and the roll failing in the future if we ever change them.)

Exposing them like this seems a bit weird because it exposes how it is nested inside the config from the root. (This was fine when it was all internal, but for exposing the JSON structure to users that seems a bit weird, non-compositional? The from and to json don't need to know how it's nested in a larger config.)

Reply:
Yes, I was actually thinking about just hard coding them everywhere, as the environment variables don't actually have to use same naming as these ones here.

Done.